### ISSUE TITLE
Potential multiple prediction

### ISSUE THREAD

**Issue by Alessiobrini** [euclidjda/deep-quant#29]
First of all, thanks for sharing this amazing repository. It is a great help for understanding those techniques applied to the finance domain.

I skimmed through the code and I was wondering if there is the possibility to improve the structure of the model and do multiple prediction at the same time. In other words I would be able to test the model on my dataset, outputting the fundamentals at t+1,t+2,t+3 for instance.

To me it looks like the model is able to provide just one single forecast at each timestep, because I don't see such a parameter among the configurations. I am just trying to understand which part of the scripts I should change in order to obtain multiple predictions.

Anyway, thank you again for your great work.

**Comment by euclidjda** [comment]
This can be done within models/deep_rnn_model.py. Or, maybe copy it and make a new model type.

At any rate, if you look at lines 139-142:

```
        if config.data_scaler is not None and config.scale_targets is True:
            self._predictions = self._reverse_center_and_scale( last_output )
        else:
            self._predictions = last_output
```
I think if you change it to:

```
        output_seq = tf.unstack(outputs, axis=1)
        if config.data_scaler is not None and config.scale_targets is True:
            self._predictions = self._reverse_center_and_scale( output_seq )
        else:
            self._predictions = output_seq
```

The unstack thingy converts outputs from tensor form to a list of tensors so that the second return value of the .step() method (implemented in models/base_model.py) will be a list of tensors, where each tensor is the output/prediction at each time step. Let me know if it works!

**Comment by euclidjda** [comment]
Actually, `self._outputs` is already a list of tensors containing the outputs so there is no need to create the `output_seq` list as I did above. But also there is an error in my solution, you can't `self._reverse_center_and_scale()` a list, you need to do it element by element.
So, something like this might be right:

```

        if config.data_scaler is not None and config.scale_targets is True:
            self._predictions = [self._reverse_center_and_scale(y) for y in self._outputs]
        else:
            self._predictions = self._outputs


```

**Comment by Alessiobrini** [comment]
I did the change you suggested and the train process seem to be ok. However when I try to predict outputs I receive the following error:

`Traceback (most recent call last):
...\predict.py", line 141, in <listcomp>
    out_str = ' '.join(["%.3f"%out[i] for i in range(len(out))])
TypeError: only size-1 arrays can be converted to Python scalars`

I suppose this is again due to the fact that now the output is a list of tensors, right? Same thing happens when I try to "pretty print" the predictions.

For what I understand (but I am totally not an expert with Tensorflow), this error could be due to the fact that we are trying to store the outputs as list of tensors, while the script expects some different datatype, as numpy arrays.

**Comment by euclidjda** [comment]
Yes, you will need to adapt predict.py to handle lists of predictions.

**Comment by Alessiobrini** [comment]
I recover this issue after a while, because maybe my changes to predict.py are not really effective. I wrote here my solution in case there will be someone like me interested in printing out multiple prediction.
I did something like the following:

```
def print_predictions(batches, batch, preds):
    key     = batch_to_key(batch)
    date    = batch_to_date(batch)

    for tens in preds:
        np.set_printoptions(suppress=True)
        np.set_printoptions(precision=3)

        out = batches.get_raw_outputs(batch,0,tens)
        flattened_out  = list(itertools.chain(*out))
        out_str = ' '.join(["%.3f"%flattened_out[i] for i in range(len(flattened_out))])

        print("%s %s %s"%(date,key,out_str))
    sys.stdout.flush()
```

Instead of taking just the first tensor from the list of tensors called preds,  the code iterates over such list of tensor, gets outputs in the raw form and then flattens the tensor to print just one row at a time. 

Using my dataset I obtain forecasts as follows (I printed just the first 10 rows):

```
     DATE     TIC  TOTAL_REVENUE_TTM  NET_INCOME_BEFORE_DISTRIBUTION_TTM  
0  201801  AMZN.O       3.769385e+11                        2.451480e+09   
1  201801  AMZN.O       3.775364e+11                        2.445514e+09   
2  201801  AMZN.O       3.772219e+11                        2.455384e+09   
3  201801  AMZN.O       3.766596e+11                        2.451482e+09   
4  201801  AMZN.O       3.757391e+11                        2.464430e+09   
5  201802  AMZN.O       3.931331e+11                        2.541014e+09   
6  201802  AMZN.O       3.935390e+11                        2.539429e+09   
7  201802  AMZN.O       3.932100e+11                        2.556680e+09   
8  201802  AMZN.O       3.926973e+11                        2.552111e+09   
9  201802  AMZN.O       3.917282e+11                        2.566256e+09
```
The date is the same every five row and I know for sure that for every group of 5 predictions, the last row represents the forecast made at the current date (so that 201801 for instance). I probably don't get what the preceding forecasts represent.

Choosing max_unrollings=5, It was natural to me expecting forecasts of shape (max_unrollings,num_variables) at every timestep for a RNN model. Anyway, I'm not sure that the way I printed the predictions is correct, because in this example I selected stride=12 and those forecast don't look to be equally spaced back in time by 12 months.

Probably the error stands in considering one tensor at a time.
